cekit-helpers/cekit-cache-server/src/main/java/org/kie/cekit/cacher/utils/CacherUtils.java

```diff
@@ -70,6 +74,49 @@ public void cleanTmpFiles() {
         }
     }
 
+    /**
+     * Clean RHPAM and RHDM nightly artifacts, keep only latest 3 nightly builds
+     * run once a day.
+     */
+    @Scheduled(every = "24h", delay = 12, delayUnit = TimeUnit.HOURS)
+    public void cleanOldProductNightlyArtifacts() throws IOException, InterruptedException {
+        log.fine("Trying to identify the latest nightly build based on filesystem artifacts to delete, this could take a while...");
+        try {
+            log.fine("Walking through data dir searching for old nightly product builds...");
+
+            Map<File, LocalDate> nightlyBuildArtifacts = new HashMap<>();
+            List<LocalDate> foundDates = new ArrayList<>();
+            Files.walk(Paths.get(cacherProperties.getCacherArtifactsDir()))
+                    .map(Path::toFile)
+                    .filter(file -> cacherProperties.buildDatePattern.matcher(file.getName()).find())
+                    .filter(file -> file.getName().endsWith(".zip"))
+                    .forEach(file -> {
+                        Matcher test = cacherProperties.buildDatePattern.matcher(file.getName());
+                        if (test.find()) {
+                            LocalDate d = LocalDate.parse(test.group(0), cacherProperties.formatter);
+                            // collect all files that matches the build date pattern
+                            nightlyBuildArtifacts.put(file, d);
+                            // collect all build dates to get the newst one to be the base date.
+                            if (!foundDates.contains(d)) {
+                                foundDates.add(d);
+                            }
+                        }
+                    });
+
+            LocalDate mostRecentNightlyBuild = Collections.max(foundDates);
+            log.fine("Latest nightly build date is -> " + mostRecentNightlyBuild);
+
+            for (Map.Entry<File, LocalDate> entry : nightlyBuildArtifacts.entrySet()) {
+                if (entry.getValue().plusDays(3).isBefore(mostRecentNightlyBuild)) {
```

Comment:
We need to look for older builds, not newer. From LocalData javadoc: "For example, 2008-12-31 plus one day would result in 2009-01-01. "
```suggestion
                if (entry.getValue().minusDays(3).isBefore(mostRecentNightlyBuild)) {
```

Reply:
not really, see that I am comparing if the current date **is before** the most recent build date found in the stored artifacts. So this sentence is correct.
suppose the latest one is 2020-10-20 and we have:
2020-10-20 + 3 days = 2020-10-23, is 2020-10-23 before than 2020-10-20?
eg:
2020-10-19 - keeps
2020-10-18 - keeps
2020-10-17 - keeps
2020-10-20 - deletes

Reply:
thanks for explanation